File: backend/app/schemas/system.py
"""System management schemas (users, roles, menus)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.common import KeywordQuery, PageQuery, SortQuery
from app.schemas.menu import MenuType

StatusFlag = Literal["enabled", "disabled"]


# ── 用户 ────────────────────────────────────────────────────────────────────
class UserItem(BaseModel):
    id: int
    username: str
    nickname: str
    email: str | None = None
    phone: str | None = None
    avatar: str | None = None
    status: StatusFlag
    roleIds: list[int] = Field(default_factory=list)
    roleNames: list[str] = Field(default_factory=list)
    lastLoginAt: str | None = None
    createdAt: str | None = None
    updatedAt: str | None = None


class UserCreateRequest(BaseModel):
    username: str
    nickname: str
    password: str
    email: str | None = None
    phone: str | None = None
    status: StatusFlag = "enabled"
    roleIds: list[int] = Field(default_factory=list)


class UserUpdateRequest(BaseModel):
    nickname: str
    email: str | None = None
    phone: str | None = None
    status: StatusFlag = "enabled"
    roleIds: list[int] = Field(default_factory=list)


class UserQueryParams(PageQuery, SortQuery, KeywordQuery):
    status: StatusFlag | None = None
    roleId: int | None = None


# ── 角色 ────────────────────────────────────────────────────────────────────
class RoleItem(BaseModel):
    id: int
    code: str
    name: str
    description: str | None = None
    status: StatusFlag
    menuIds: list[int] = Field(default_factory=list)
    createdAt: str | None = None
    updatedAt: str | None = None


class RoleCreateRequest(BaseModel):
    code: str
    name: str
    description: str | None = None
    status: StatusFlag = "enabled"


class RoleUpdateRequest(RoleCreateRequest):
    pass


class RoleMenuAssignment(BaseModel):
    roleId: int
    menuIds: list[int] = Field(default_factory=list)


class RoleQueryParams(PageQuery, SortQuery, KeywordQuery):
    status: StatusFlag | None = None


# ── 菜单 ────────────────────────────────────────────────────────────────────
class MenuCreateRequest(BaseModel):
    menuCode: str
    parentId: int | None = None
    type: MenuType
    title: str
    path: str | None = None
    componentPath: str | None = None
    icon: str | None = None
    permissionCode: str | None = None
    order: int = 0
    visible: bool = True
    enabled: bool = True
    tagViewEnabled: bool = True
    keepAliveEnabled: bool = False


class MenuUpdateRequest(MenuCreateRequest):
    pass


class MenuQueryParams(KeywordQuery):
    type: MenuType | None = None
    enabled: bool | None = None
    visible: bool | None = None

File: frontend/src/types/system.ts
import type { Id, KeywordQuery, PageQuery, SortQuery, StatusFlag } from './common';
import type { MenuType } from './menu';

export interface UserItem {
  id: Id;
  username: string;
  nickname: string;
  email?: string | null;
  phone?: string | null;
  avatar?: string | null;
  status: StatusFlag;
  roleIds: Id[];
  roleNames?: string[];
  lastLoginAt?: string | null;
  createdAt?: string;
  updatedAt?: string;
}

export interface UserQueryParams extends PageQuery, SortQuery, KeywordQuery {
  status?: StatusFlag;
  roleId?: Id;
}

export interface UserCreateRequest {
  username: string;
  nickname: string;
  password: string;
  email?: string | null;
  phone?: string | null;
  status: StatusFlag;
  roleIds: Id[];
}

export interface UserUpdateRequest {
  nickname: string;
  email?: string | null;
  phone?: string | null;
  status: StatusFlag;
  roleIds: Id[];
}

export interface RoleItem {
  id: Id;
  code: string;
  name: string;
  description?: string | null;
  status: StatusFlag;
  menuIds: Id[];
  createdAt?: string;
  updatedAt?: string;
}

export interface RoleQueryParams extends PageQuery, SortQuery, KeywordQuery {
  status?: StatusFlag;
}

export interface RoleCreateRequest {
  code: string;
  name: string;
  description?: string | null;
  status: StatusFlag;
}

export type RoleUpdateRequest = RoleCreateRequest;

export interface RoleMenuAssignment {
  roleId: Id;
  menuIds: Id[];
}

export interface SystemMenuItem {
  id: Id;
  menuCode: string;
  parentId?: Id | null;
  type: MenuType;
  title: string;
  path?: string | null;
  componentPath?: string | null;
  icon?: string | null;
  permissionCode?: string | null;
  order: number;
  visible: boolean;
  enabled: boolean;
  tagViewEnabled: boolean;
  keepAliveEnabled: boolean;
  children: SystemMenuItem[];
  createdAt?: string;
  updatedAt?: string;
}

export interface MenuQueryParams extends KeywordQuery {
  type?: MenuType;
  enabled?: boolean;
  visible?: boolean;
}

export interface MenuCreateRequest {
  menuCode: string;
  parentId?: Id | null;
  type: MenuType;
  title: string;
  path?: string | null;
  componentPath?: string | null;
  icon?: string | null;
  permissionCode?: string | null;
  order?: number;
  visible?: boolean;
  enabled?: boolean;
  tagViewEnabled?: boolean;
  keepAliveEnabled?: boolean;
}

export type MenuUpdateRequest = MenuCreateRequest;
